v0.5/fastn-lang/src/f-script/README.md

```diff
@@ -0,0 +1,221 @@
+# F-script appearances
+
+- body of a function definition
+
+```ftd
+-- string foo():
+string name:
+
+if name == "" {
+  "<default>"
+} else {
+  name
+}
+```
+
+The body contains a list of expressions. The last expression is evaluated and
```

Comment:
We should introduce the new terminology. Not everything is an expression. 

````ftd
-- string foo():
string name:

string $foo: "hello";

if name.len() > 4 {
    foo = "bar";
}

if foo > name { foo } else { name }
```

The first and second statements in this function are not expression. Expressions are required only in the return position of a block, which is either an explicit return statement, or the last statement of the block.

Reply:
I don't think we gain anything by making this distinction. Even in rust, `if` is not a statement but an expression. 

```rust
if name.len() > 4 {
    foo = "bar";
}
```

This `if` expression in rust would evaluate to `()` so this will compile.

```rust
if name.len() > 4 {
    4
} else {
    3
}
```

The above won't compile, the error would look something like:

```
error[E0308]: mismatched types
  --> test.rs:9:9
   |
8  | /     if 2 > 4 {
9  | |         4
   | |         ^ expected `()`, found integer
10 | |     } else {
11 | |         3
12 | |     }
   | |_____- expected this to be `()
```

Variable declarations can even be treated like expression that evaluate to `()` so you don't have to use the result of this expression. Evaluation of this expression would introduce a new symbol in your scope.


```rust
if foo > name { foo } else { name }
```

The above will either evaluate to `foo` or `name`. Since this is the last expression, it's value will be returned.

`match` and `for` can be thought of as expressions too. I am not trying to introduce "statement" unless I have to.

Reply:
Makes sense.